keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md

```diff
@@ -0,0 +1,832 @@
+- Workload controllers
+- Workload
+- Default
+- Policies
+
+Each resource type will be assigned a  `NamespaceDeletionOrder` value(<TOBESOLVED - @cici37>int value or string value). To define the DeletionOrder, the options are:
+
+1. Add a field into APIResource (not mutable) so it could be observed easily(pkg/registry):
+
+```
+type APIResource struct {
+……
+`NamespaceDeletionOrder int64 `json:"namespaceDeletionOrder" protobuf:"varint,11,opt,name=namespaceDeletionOrder"``
+}
+```
+
+In this case, the `NamespaceDeletionOrder` will be associated with the resources naturally and the resource which does not have an opinion will default to a value.
+
+- Pros:
+  - Having it configurable
+  - Consistent between CRD and native type when add support into CRD
+- Cons:
+  - Hard to oversee the overall prioritization across resources
+
+2. Maintain a hard-coded map of DeletionOrder per resources for the resources which have a deletion order preference. Any resources which have no preference opinion would be in the Default category.
+```
+var resourceDeletionOrderPriorities = map[ResourceType]DeletionOrderPriority{
+	"/pods": DeletionOrdeWorkload,
+	"apps.k8s.io/deployments":  DeletionOrdeWorkloadController.
+	"networking.k8s.io/networkpolicies”: DeletionOrdePolicy,
+	……
+})
+
+```
+
+- Pros:
+  - The feature would be fully under control
+  - Have one single place to manage the DeletionOrderPriority across all resources group
+  - Easy for future maintenance
+- Cons:
+  - Not configurable
+  - Not encapsulated (behavior of a type/group defined far away from that type’s “main” code)
+  - Only apply to native types, need separate support for CRD
+
+
+
+During namespace deletion, the namespace collector will traverse resources in the namespace and delete them in ascending priority order based on the `NamespaceDeletionOrder` defined.
+
+### Handling Cyclic Dependencies
+
+Cyclic dependencies can occur if resources within the namespace have finalizers set which conflicts with the DeletionOrderPriority. 
+For example, consider the following scenario:
+
+- Resource A has a finalizer that depends on the deletion of Resource B.
+
+- Resource A is in the earlier DeletionOrderPriority than Resource B.
+
+In this case, the finalizers set would conflict with the `NamespaceDeletionOrder` set for resources and could lead to cyclic dependencies and cause namespace deletion process hanging.
+
+To address this, the system will:
+
+- Attempt to honor the DeletionOrderPriority for resource deletion.
+
+- Monitor the deletion process for each `NamespaceDeletionOrder`. If the process hangs beyond a predefined timeout (e.g., 5 minutes), 
+  it will detect the stall and trigger the deletion attempt for the next `NamespaceDeletionOrder` group.
+
+- If the resources from the next `NamespaceDeletionOrder` group are deleted successfully, retry with the ordered namespace deletion mechanism. If the process hangs again, wait for a predefined timeout and retry.
```

Comment:
Is there anything to "retry"?

Assume your order says all A's before B's before C's.  You delete all A objects, but at least one is still hanging around.  You hit the timeout and move on to B.  You delete all B objects.  Now you are waiting for all A || B objects.  If you hit the timeout, you move onto C.

Right?

Reply:
Rephrased it.